### PR TITLE
Implement GenesisCoconsensus

### LINKED ISSUE #196: Implement GenesisCoconsensus contract
1. Implement `GenesisCoconsensus`
    - It has a mapping `genesisMetachainIds`
    - For now, this will have just one value in the mapping
    - The tracking of MetachainIds should be such that it can be iterable

### PR DESCRIPTION
**Pull request by deepesh-kn**
This pull request contains the implementation of `GenesisCoconsensus` contract.
fixes: #196

### CODE REVIEW

File: contracts/coconsensus/GenesisCoconsensus.sol [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// limitations under the License.
+
+/**
+ * @title Genesis coconsensus contract is a storage contract that holds
+ *        the initial values required by the contract that were written in the
+ *        genesis block. This contract stores the information related to metachains.
+ *        Coconsensus can track multiple protocores, so the values needed to
+ *        initialize protocore is stored in the mapping. The metachain id's are
+ *        stored as a linked list and is iterable. The corresponding core address,
+ *        epoch length and metablock height can be retrived for a given metachain
```
Comment (jayeshbairagi) [inline contracts/coconsensus/GenesisCoconsensus.sol]:
```suggestion
 *        epoch length and metablock height can be retrieved for a given metachain
```

File: contracts/coconsensus/GenesisCoconsensus.sol [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    /** Link list of metablock ids (for other protocores).  */
+    mapping(bytes32 => bytes32) public genesisMetachainIds;
+
+    /** Mapping of metablock id to its core contract address on origin chain. */
+    mapping(bytes32 => address) public genesisCores;
+
+    /** Mapping of metablock id to the epoch lengths. */
+    mapping(bytes32 => uint256) public genesisEpochlengths;
+
+    /** Mapping of metablock id to the metblock height. */
```
Comment (jayeshbairagi) [inline contracts/coconsensus/GenesisCoconsensus.sol]:
```suggestion
    /** Mapping of metablock id to the metablock height. */
```

File: contracts/coconsensus/GenesisCoconsensus.sol [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    bytes32 public genesisMetachainId;
+
+    /** Link list of metablock ids (for other protocores).  */
+    mapping(bytes32 => bytes32) public genesisMetachainIds;
+
+    /** Mapping of metablock id to its core contract address on origin chain. */
+    mapping(bytes32 => address) public genesisCores;
+
+    /** Mapping of metablock id to the epoch lengths. */
+    mapping(bytes32 => uint256) public genesisEpochlengths;
```
Comment (jayeshbairagi) [inline contracts/coconsensus/GenesisCoconsensus.sol]:
```suggestion
    mapping(bytes32 => uint256) public genesisEpochLengths;
```

File: contracts/coconsensus/GenesisCoconsensus.sol [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @title Genesis coconsensus contract is a storage contract that holds
+ *        the initial values required by the contract that were written in the
+ *        genesis block. This contract stores the information related to metachains.
+ *        Coconsensus can track multiple protocores, so the values needed to
+ *        initialize protocore is stored in the mapping. The metachain id's are
```
Comment (jayeshbairagi) [inline contracts/coconsensus/GenesisCoconsensus.sol]:
```suggestion
 *        initialize protocore are stored in the mapping. The metachain id's are
```

**Review by jayeshbairagi — CHANGES_REQUESTED** [review]
Minor typos

File: contracts/coconsensus/GenesisCoconsensus.sol (hunk 1 above)
Comment (deepesh-kn) [reply contracts/coconsensus/GenesisCoconsensus.sol]: done

File: contracts/coconsensus/GenesisCoconsensus.sol (hunk 2 above)
Comment (deepesh-kn) [reply contracts/coconsensus/GenesisCoconsensus.sol]: done

File: contracts/coconsensus/GenesisCoconsensus.sol (hunk 3 above)
Comment (deepesh-kn) [reply contracts/coconsensus/GenesisCoconsensus.sol]: done

File: contracts/coconsensus/GenesisCoconsensus.sol (hunk 4 above)
Comment (deepesh-kn) [reply contracts/coconsensus/GenesisCoconsensus.sol]: done

**Review by jayeshbairagi — APPROVED** [review]
LGTM

**Pull request merged**